### ISSUE TITLE
`collection_methods_unrelated_type` false positive with `MapMixin`

### ISSUE THREAD

**Issue by oprypin** [dart-lang/linter#4004]
I found real code that's equivalent to the below, where `collection_methods_unrelated_type` triggers. As I understand, that should mean that the map lookup can never succeed, however, here is an example where it clearly succeeds.

```dart
import 'dart:collection';

enum Field { foo }

class TestClass with MapMixin<String, Object?> {
  int foo = 1234;

  @override
  Object? operator [](Object? key) {
    switch (key) {
      case 'foo':
        return foo;
      case Field.foo:
        return foo;
    }
    return null;
  }

  // Irrelevant, minimal implementations:
  @override
  operator []=(Object? key, Object? value) {}
  @override
  List<String> get keys => const ['foo'];
  @override
  void clear() {}
  @override
  void remove(Object? key) {}
}

void main() {
  final obj = TestClass();
  final x = obj[Field.foo]; // LINT: 'Field' isn't related to 'String'.
  print(x); // 1234, which should be impossible according to the lint.
}
```

Of course, the issue is that the class says that it accepts `String` keys but instead implements a much wider override, so one could argue that maybe this isn't even a false positive. But I still wanted to report this, because blindly fixing the lint was dangerous in this case.

**Comment by srawlins** [comment]
Thanks for reporting because I did not even conceive of this possibility 🤣 . I think this is pretty fascinating.

I'm inclined to say, 'Bad Mixin!' and say the lint is functioning correctly. But maybe we can be more precise about what Map types are relevant.

I imagine @lrhn and @bwilkerson and @pq would be interested in this.

**Comment by lrhn** [comment]
You can do lots of silly things, and if you do so, we can't really help you.
(If anyone ever created a `Future` which calls its callbacks more than once, or both error and value callbacks, or a stream that calls listeners while its paused or after cancel, I have absolutely no idea what will happen. I hope that at least our `async` functions will fail early.)

Anyone writing code like the one above should fail their code-review and be sent back to the drawing board.
It's a mock map implementation for testing, which behaves in a way that no real map ever will, so it's not clear what it's supposed to be testing.

If an object implements `Map`, it's supposed to follow the `Map` protocol, which means representing a *finite* map from keys to values. Any key not mentioned in `map.keys` is not supposed to return a non-`null` value when you look it up. The specification of map patterns is written around that assumption &mdash; carefully, so it doesn't do something inconsistent or unsound if the assumption fails to be true, but still assuming that violations won't happen for code that we care about.

Warning about the code here is perfectly reasonable, and *not* a false positive.
If that's a problem for the authors, they'll have to add `// ignore` comments.

**Comment by oprypin** [comment]
> It's a mock map implementation for testing

I am not sure what this is referring to. If you're talking about my example being named `TestClass`, that's just because it's a heavily modified example. The actual code I found was not in testing.

**Comment by lrhn** [comment]
> The actual code I found was not in testing.

That's scary!

My conclusion stands &mdash; if someone wants to make a map that behaves in a non-`Map`-like way in production code, that's entirely on the author, and not something the lint should cater to. The lint is valid. The code that fails it is highly non-idiomatic, bordering on questionable, and should be accompanied by copious amounts of documentation. Having to add an `// ignore` line is reasonable.

**Comment by pq** [comment]
I'm with Lasse.  Pathological cases like these are exactly what ignore comments are for.  If the ignores offend, it's a good signal to refactor!
